python/lsst/ip/isr/deferredCharge.py

```diff
@@ -0,0 +1,639 @@
+
+        This implements equation 10 of Snyder+21.  For an image with
+        CTI, s'(m, n), the correction factor is equal to the maximum
+        value of the set of:
+            {A_L s'(m, n - j) exp(-j t / \tau_L)}_j=0^jmax
+
+        Parameters
+        ----------
+        inputArr : `np.ndarray`, (nx, ny)
+            Input image data to correct.
+        drift_scale : `float`
+            Drift scale (Snyder+21 A_L value) to use in correction.
+        decay_time : `float`
+            Decay time (Snyder+21 \tau_L) of the correction.
+        num_previous_pixels : `int`, optional
+            Number of previous pixels to use for correction.  As the
+            CTI has an exponential decay, this essentially truncates
+            the correction where that decay scales the input charge to
+            near zero.
+
+        Returns
+        -------
+        outputArr : `np.ndarray`, (nx, ny)
+            Corrected image data.
+        """
+        r = np.exp(-1/decay_time)
+        Ny, Nx = inputArr.shape
+
+        # j = 0 term:
+        offset = np.zeros((num_previous_pixels, Ny, Nx))
+        offset[0, :, :] = drift_scale*np.maximum(0, inputArr)
+
+        # j = 1..jmax terms:
+        for n in range(1, num_previous_pixels):
+            offset[n, :, n:] = drift_scale*np.maximum(0, inputArr[:, :-n])*(r**n)
+
+        Linv = np.amax(offset, axis=0)
+        outputArr = inputArr - Linv
+
+        return outputArr
+
+    @staticmethod
+    def local_trap_inverse(inputArr, trap, global_cti=0.0, num_previous_pixels=6):
+        """Apply localized trapping inverse operator to pixel signals.
+
+        This implements equation 13 of Snyder+21.  For an image with
+        CTI, s'(m, n), the correction factor is equal to the maximum
+        value of the set of:
+            {A_L s'(m, n - j) exp(-j t / \tau_L)}_j=0^jmax
+
+        Parameters
+        ----------
+        inputArr : `np.ndarray`, (nx, ny)
+            Input image data to correct.
+        trap : `lsst.ip.isr.SerialTrap`
+            Serial trap describing the capture and release of charge.
+        global_cti: `float`
+            Mean charge transfer inefficiency, b from Snyder+21.
+        num_previous_pixels : `int`, optional
+            Number of previous pixels to use for correction.
+
+        Returns
+        -------
+        outputArr : `np.ndarray`, (nx, ny)
+            Corrected image data.
+
+        """
+        Ny, Nx = inputArr.shape
+        a = 1 - global_cti
+        r = np.exp(-1/trap.emission_time)
+
+        # Estimate trap occupancies during readout
+        trap_occupancy = np.zeros((num_previous_pixels, Ny, Nx))
+        for n in range(num_previous_pixels):
+            trap_occupancy[n, :, n+1:] = trap.capture(np.maximum(0, inputArr))[:, :-(n+1)]*(r**n)
+        trap_occupancy = np.amax(trap_occupancy, axis=0)
+
+        # Estimate captured charge
+        C = trap.capture(np.maximum(0, inputArr)) - trap_occupancy*r
+        C[C < 0] = 0.
```

Comment:
I think previously you've used `np.clip()` for this. I wonder if that's any better here or not (probably doesn't matter as the array is small though).